ucp/application_context.py

```diff
@@ -0,0 +1,376 @@
+        if ctx is None:
+            return
+        ctx.progress()
+        del ctx
+        await asyncio.sleep(0)
+
+
+async def _arm_worker(weakref_ctx, rsock, event_loop):
+    """This help function args the worker.
+    Notice, it only keeps a weak reference to `ApplicationContext`, which makes it
+    possible to call `ucp.reset()` even when this loop is running.
+    """
+
+    # When arming the worker, the following must be true:
+    #  - No more progress in UCX (see doc of ucp_worker_arm())
+    #  - All asyncio tasks that isn't waiting on UCX must be executed
+    #    so that the asyncio's next state is epoll wait.
+    #    See <https://github.com/rapidsai/ucx-py/issues/413>
+    while True:
+        ctx = weakref_ctx()
+        if ctx is None:
+            return
+        ctx.progress()
+        # This IO task returns when all non-IO tasks are finished.
+        await event_loop.sock_recv(rsock, 1)
+        if ctx._worker.arm():
+            # At this point we know that asyncio's next state is
+            # epoll wait.
+            break
+        del ctx
+
+
+class ApplicationContext:
+    def __init__(self, config_dict={}, blocking_progress_mode=None):
+        self.event_loops_binded_for_progress = set()
+        self.progress_tasks = []
+        self.dangling_arm_task = None
+        self.epoll_fd = -1
+        self._ctx = None
+        self._worker = None
+        if blocking_progress_mode is not None:
+            self.blocking_progress_mode = blocking_progress_mode
+        elif "UCXPY_NON_BLOCKING_MODE" in environ:
+            self.blocking_progress_mode = False
+        else:
+            self.blocking_progress_mode = True
+
+        # Create context and worker, whic might fail
+        self._ctx = ucx_api.UCXContext(config_dict)
+        self._worker = ucx_api.UCXWorker(self._ctx)
+
+        if self.blocking_progress_mode:
+            self.epoll_fd = self._worker.init_blocking_progress_mode()
+
+    def __del__(self):
+        for task in self.progress_tasks:
+            task.cancel()
+        if self.dangling_arm_task is not None:
+            self.dangling_arm_task.cancel()
+        # Notice, worker and context might never have been created
+        if self._worker is not None:
+            self._worker.destroy()
+        if self._ctx is not None:
+            self._ctx.destroy()
+        if self.blocking_progress_mode and self.epoll_fd != -1:
+            close_fd(self.epoll_fd)
+
+    def create_listener(self, callback_func, port, guarantee_msg_order):
+        self.continuous_ucx_progress()
+        if port in (None, 0):
+            # Get a random port number and check if it's not used yet. Doing this
+            # without relying on `socket` allows preventing UCX errors such as
+            # "none of the available transports can listen for connections", due
+            # to the socket still being in TIME_WAIT state.
+            try:
+                with open("/proc/sys/net/ipv4/ip_local_port_range") as f:
```

Comment:
This feels a little fragile just because there may be systems that don't have this. However given your point above, I'm not sure of another option. Just noting this concern for now.

Reply:
This is not really part of this PR, but was added by me in https://github.com/rapidsai/ucx-py/pull/423. For systems that don't have this, we have default values in the `except` block.

Reply:
Ok I'll raise an issue then.